### PR TITLE
WIP: Make EVP_PKEY_asn1_add0 thread safe

### PR DESCRIPTION
**Pull request by levitte**

### CODE REVIEW

**Comment by levitte** [comment]
I was considering backporting to 1.0.2 as well....  but the locking API is damn different there, so I'm hesitant.

**Comment by bernd-edlinger** [comment]
But if we go that way, the find method needs a read lock.

**Comment by bernd-edlinger** [comment]
For the "use case" I have in mind so far, it would also be sufficient to initialize
the app_methods before the multi-threaded application actually starts.

**Comment by levitte** [comment]
Ah, good point.  Getting to it

**Comment by levitte** [comment]
Regarding initialising `app_methods` earlier, I'm not sure I see a reason why.  If the application isn't adding any `EVP_PKEY_ASN1_METHOD`, why do we need to allocate that stack?

**Comment by bernd-edlinger** [comment]
Aehm, sorry.
I meant if all the asn1_add0 method calls happen at app start up time
before the first thread is started, the code could as well stay as it is
and just be documented that way. 
OTOH if I have an API like EVP_PKEY_asn1_count() even if it is
internally thread-safe what does it mean when it is used in a loop termination?
and EVP_PKEY_asn1_get0 even if it is internally thread safe,
will that work reasonable, when in another thread a new app_method
is inserted, and sorted?

**Comment by bernd-edlinger** [comment]
I mean code like this:

```
    for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
        ameth = EVP_PKEY_asn1_get0(i);
```

needs to lock the app_mutex over the whole loop.
and by the way are OpenSSL mutexes recursive?

**Comment by levitte** [comment]
Gotcha.  I've added a new commit that basically read locks everywhere `app_methods` is looked at.

**Comment by bernd-edlinger** [comment]
Maybe when the sort is stable, which means the existing entries in the app_methods
keep their relative positions then you can get away with the race condition in the above loop.
But if the loop traverses the entries in reverse order it can fail to look up an
existing entry which is moved to higher position while the loop executes:
```
for (i = EVP_PKEY_asn1_get_count()-1; i>=0 i--) {
        ameth = EVP_PKEY_asn1_get0(i);
```

**Comment by levitte** [comment]
Sorry, misunderstood you earlier.  Does my latest commit appear sane enough?

File: crypto/asn1/ameth_lib.c [hunk 1]
```diff
@@ -183,13 +188,23 @@ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find_str(ENGINE **pe,
...
-            && (strncasecmp(ameth->pem_str, str, len) == 0))
-            return ameth;
+
+    /*
+     * Just as in EVP_PKEY_asn1_get_count(), if we can't init the app_methods
+     * lock, we pretend that there are no app_methods.
+     */
+    if (RUN_ONCE(&app_methods_init, do_app_methods_init)) {
+        CRYPTO_THREAD_read_lock(app_methods_lock);
+        for (i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
```
Comment (bernd-edlinger) [inline crypto/asn1/ameth_lib.c]: This creates a recursive read_lock.  I don't know if that is safe.

**Comment by bernd-edlinger** [comment]
I am not sure if only the implementation needs to be made thread-safe.
This is probably an example of an _interface_ that is inherently thread-unsafe.

**Comment by levitte** [comment]
You're right about that...  or rather, not all of it, the finders are safe to use, but it's the combination of `EVP_PKEY_asn1_get_count` and `EVP_PKEY_asn1_get0` that isn't.  In other words, there's a need for a thread safe iterator (and that's where the thought of a context comes in, right?).

**Comment by bernd-edlinger** [comment]
Yes, having a context to store such things would help a lot.
Have you ever thought of removing a EVP_PKEY_ASN1_METHOD again?
Say if you have a shared object/DLL that needs to be unloaded again?

**Comment by levitte** [comment]
I'm not sure how a context to store these things would help.  They are used internally in all sorts of places, so that's hardly feasible.  However, having a temporary context that sort of "locks" the database (or rather, takes a snapshot) any time someone wants to search through it, *that* would be feasible.

Removal of a `EVP_PKEY_ASN1_METHOD` isn't currently part of the API, and I think that's subject of a different PR altogether, if ever.

**Comment by bernd-edlinger** [comment]
I am OK with your suggestion.
Also just requiring the application to do the app_methods initialization before the
first use would be acceptable, which means just documenting how it works now.

**Comment by levitte** [comment]
Regarding this:

> I am not sure if only the implementation needs to be made thread-safe.
> This is probably an example of an _interface_ that is inherently thread-unsafe.

... I just re-read [attributes(7)](http://man7.org/linux/man-pages/man7/attributes.7.html), because there are definitions in there that are applicable for this case, and most of all, what thread safe actually means:

```
       MT-Safe
              MT-Safe or Thread-Safe functions are safe to call in the
              presence of other threads.  MT, in MT-Safe, stands for Multi
              Thread.

              Being MT-Safe does not imply a function is atomic, nor that it
              uses any of the memory synchronization mechanisms POSIX
              exposes to users.  It is even possible that calling MT-Safe
              functions in sequence does not yield an MT-Safe combination.
              For example, having a thread call two MT-Safe functions one
              right after the other does not guarantee behavior equivalent
              to atomic execution of a combination of both functions, since
              concurrent calls in other threads may interfere in a
              destructive way.

              Whole-program optimizations that could inline functions across
              library interfaces may expose unsafe reordering, and so
              performing inlining across the GNU C Library interface is not
              recommended.  The documented MT-Safety status is not
              guaranteed under whole-program optimization.  However,
              functions defined in user-visible headers are designed to be
              safe for inlining.
```

Here, this is key: *It is even possible that calling MT-Safe functions in sequence does not yield an MT-Safe combination.*

**Comment by levitte** [comment]
Hey...  haven't closed this one yet...  would you say it's worth the while keeping this for master if we add an iterator of the sort I mentioned?

**Comment by bernd-edlinger** [comment]
Hmmmmmmmmm......

trouble is how to test that.  If you had a thread that adds and removes stuff
from the table, while another thread could be affected by a race, that would
be a realistic test case.

But in the moment we can just add stuff, until the table is saturated, this feels
like a car with 3 wheels...

OTOH if an application does not register all ASN1 stuff before the first
use then something will fail even if there is no race condition, just because
something is missing in the table when it is needed.

Maybe, there should be some way for the application to register
the ASN1 objects in a kind of init_once function, maybe something like
a weak symbol.  This is of course only a vague idea...

**Comment by mattcaswell** [comment]
@levitte what is the status of this PR?

**Comment by richsalz** [comment]
This is not a "just in 1.1.1" issue, so removing that milestone.

**Comment by t8m** [comment]
I suppose this is obsolete now and can be closed?

**Comment by t8m** [comment]
@levitte ?

**Comment by t8m** [comment]
Closing as EVP_PKEY_asn1 APIs are legacy.

**Pull request closed without merging**